vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/PriorityByWorkflow.java

```diff
@@ -117,29 +101,31 @@ public synchronized ConsumableResourceResponse request(
               + "values should be one of the following: %s", workflowName, workflowPriority,
               acceptedPriorities.stream().map(String::valueOf).collect(Collectors.joining(", "))));
     }
+    if (workflows.get(workflowName) == null) {
+      set(workflowName, vidarrId, input);
+    }
+
     final var state = workflows.get(workflowName);
-    if (state == null) {
-      return ConsumableResourceResponse.error(
-          String.format("Internal Vidarr error: The %s workflow run priority has not been configured properly.", workflowName));
+
+    if (state == null || state.waiting.isEmpty()) {
+      return ConsumableResourceResponse.AVAILABLE;
+    }
+    if (workflowPriority >= state.waiting.last().getValue()) {
+      return ConsumableResourceResponse.AVAILABLE;
     } else {
-      if (workflowPriority >= state.waiting.last().getValue()) {
-        state.waiting.remove(new SimpleEntry(vidarrId, workflowPriority));
-        currentInPriorityWaitingCount.labels(workflowName).set(state.waiting.size());
-        return ConsumableResourceResponse.AVAILABLE;
-      } else {
-        state.waiting.add(new SimpleEntry(vidarrId, workflowPriority));
-        currentInPriorityWaitingCount.labels(workflowName).set(state.waiting.size());
-        return ConsumableResourceResponse.error(
-            String.format("There are %s workflows currently queued up with higher priority.", workflowName));
-      }
+      set(workflowName, vidarrId, input);
+      return ConsumableResourceResponse.error(
+          String.format("There are %s workflows currently queued up with higher priority.",
+              workflowName));
     }
+
   }
 
-  public void set(String workflowName, String vidarrId, JsonNode input) {
+  public void set(String workflowName, String vidarrId, Optional<JsonNode> input) {
```

Comment:
is this the same logic as recover()?

Reply:
... yes
I could essentially replace it with a call to set

Reply:
Hmm, good point - and thinking through the `recover` logic, I think we've got it backwards here. For most consumable resources, they lay a claim on the resource while the workflow run is running - "I'm claiming this max-in-flight slot, I'm claiming this much RAM". PriorityByWorkflow is a bit different - once the workflow run has gone through the check to determine it's the best one to launch, it doesn't "hold on to" the resource (once the workflow run launches, it is no longer tracked in PriorityByWorkflow). [`recover` gets called for workflows that are out of "WAITING_FOR_RESOURCES"](https://github.com/oicr-gsi/vidarr/blob/master/vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/DatabaseBackedProcessor.java#L743), ie. launched, so I think the logic for PriorityByWorkflow's `recover` should be to do nothing.

Is that also your read of how the function works?

Reply:
Ok taking a look at where it's called that makes sense